sdk/elixir/runtime/main.go

```diff
@@ -0,0 +1,177 @@
+import (
+	"context"
+	"fmt"
+	"path"
+
+	"github.com/iancoleman/strcase"
+)
+
+const (
+	ModSourceDirPath     = "/src"
+	sdkSrc               = "/sdk"
+	genDir               = "sdk"
+	schemaPath           = "/schema.json"
+	defaultElixirVersion = "1.16.1-erlang-26.2.2-debian-bookworm-20240130-slim"
+)
+
+func New(
+	// +optional
+	sdkSourceDir *Directory,
+) *ElixirSdk {
+	return &ElixirSdk{SDKSourceDir: sdkSourceDir, RequiredPaths: []string{}}
+}
+
+type ElixirSdk struct {
+	SDKSourceDir  *Directory
+	RequiredPaths []string
+}
+
+func (m *ElixirSdk) ModuleRuntime(
+	ctx context.Context,
+	modSource *ModuleSource,
+	introspectionJson string,
+) (*Container, error) {
+	ctr, err := m.CodegenBase(ctx, modSource, introspectionJson)
+	if err != nil {
+		return nil, err
+	}
+
+	modName, err := modSource.ModuleName(ctx)
+	if err != nil {
+		return nil, err
+	}
+	subPath, err := modSource.SourceSubpath(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	mod := normalizeModName(modName)
+	entrypoint := path.Join(ModSourceDirPath, subPath, mod)
+
+	return ctr.
+		WithEntrypoint([]string{
+			"mix", "cmd",
+			"--cd", entrypoint,
+			"mix do deps.get + dagger.invoke",
+		}), nil
+}
+
+func (m *ElixirSdk) Codegen(
+	ctx context.Context,
+	modSource *ModuleSource,
+	introspectionJson string,
+) (*GeneratedCode, error) {
+	ctr, err := m.CodegenBase(ctx, modSource, introspectionJson)
+	if err != nil {
+		return nil, fmt.Errorf("could not load module config: %v", err)
+	}
+
+	return dag.GeneratedCode(ctr.Directory(ModSourceDirPath)).
+		WithVCSGeneratedPaths([]string{genDir + "/**"}).
+		WithVCSIgnoredPaths([]string{"dagger"}), nil
+}
+
+func (m *ElixirSdk) CodegenBase(
+	ctx context.Context,
+	modSource *ModuleSource,
+	introspectionJson string,
+) (*Container, error) {
+	modName, err := modSource.ModuleName(ctx)
+	if err != nil {
+		return nil, err
+	}
+	subPath, err := modSource.SourceSubpath(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	mod := normalizeModName(modName)
+
+	codegenDepsCache, codegenBuildCache := mixProjectCaches(dag, "dagger-codegen")
+
+	ctr := m.Base("").
+		WithMountedDirectory(ModSourceDirPath, modSource.ContextDirectory()).
+		WithMountedDirectory(sdkSrc, m.SDKSourceDir).
+		WithMountedCache(codegenPath()+"/deps", codegenDepsCache).
+		WithMountedCache(codegenPath()+"/_build", codegenBuildCache).
+		With(installCodegen).
+		WithNewFile(schemaPath, ContainerWithNewFileOpts{
+			Contents: introspectionJson,
+		}).
+		WithWorkdir(path.Join(ModSourceDirPath, subPath)).
+		WithDirectory(
+			"dagger",
+			m.SDKSourceDir,
+			ContainerWithDirectoryOpts{Exclude: []string{
+				"*.livemd",
+				"*.md",
+				".changes",
+				"dagger_codegen",
+				"runtime",
+				"scripts",
+				"test",
+			}},
+		).
+		WithWorkdir(path.Join(ModSourceDirPath, subPath, "dagger")).
+		WithExec([]string{
+			"dagger_codegen", "generate",
+			"--outdir", "lib/dagger/gen",
+			"--introspection", schemaPath,
+		}).
+		WithExec([]string{
+			"mix", "format",
+		}).
```

Comment:
If this takes more than ms, it can be skipped.

Reply:
1s exactly. T_T

<img width="1658" alt="Screenshot 2024-04-21 164344" src="https://github.com/dagger/dagger/assets/484530/cfcc44ad-9eec-4fe3-9405-81d36cc9c4de">

Reply:
I temporary removing it for now to make module startup faster.